src/util-exception-policy.c

```diff
@@ -61,14 +65,73 @@ void SetMasterExceptionPolicy(void)
     g_eps_master_switch = ExceptionPolicyParse("exception-policy", true);
 }
 
-static enum ExceptionPolicy GetMasterExceptionPolicy(void)
+enum ExceptionPolicy GetMasterExceptionPolicy(void)
 {
     return g_eps_master_switch;
 }
 
+static uint16_t ExceptionPolicyFlag(enum PacketDropReason drop_reason)
+{
+    switch (drop_reason) {
+        case PKT_DROP_REASON_DEFRAG_MEMCAP:
+            return EXCEPTION_DEFRAG_MEMCAP;
+        case PKT_DROP_REASON_STREAM_MEMCAP:
+            return EXCEPTION_SESSION_MEMCAP;
+        case PKT_DROP_REASON_STREAM_REASSEMBLY:
+            return EXCEPTION_REASSEMBLY_MEMCAP;
+        case PKT_DROP_REASON_FLOW_MEMCAP:
+            return EXCEPTION_FLOW_MEMCAP;
+        case PKT_DROP_REASON_STREAM_MIDSTREAM:
+            return EXCEPTION_MIDSTREAM;
+        case PKT_DROP_REASON_APPLAYER_ERROR:
+            return EXCEPTION_APPLAYER_ERROR;
+        default:
+            return BIT_U16(0);
+    }
+    return BIT_U16(0);
+}
+
+const char *ExceptionPolicyTargetFlagToString(uint16_t target_flag)
+{
+    if (target_flag & EXCEPTION_DEFRAG_MEMCAP)
+        return "defrag_memcap";
+    if (target_flag & EXCEPTION_SESSION_MEMCAP)
+        return "stream_memcap";
+    if (target_flag & EXCEPTION_REASSEMBLY_MEMCAP)
+        return "stream_reassembly_memcap";
+    if (target_flag & EXCEPTION_FLOW_MEMCAP)
+        return "flow_memcap";
+    if (target_flag & EXCEPTION_MIDSTREAM)
+        return "stream_midstream";
+    if (target_flag & EXCEPTION_APPLAYER_ERROR)
+        return "app_layer_error";
+    return "ignore";
+}
+
+enum ExceptionPolicy ExceptionPolicyTargetPolicy(uint16_t target_flag)
+{
+    if (target_flag & EXCEPTION_DEFRAG_MEMCAP)
+        return DefragGetMemcapExceptionPolicy();
+    if (target_flag & EXCEPTION_SESSION_MEMCAP)
+        return StreamTcpSsnMemcapGetExceptionPolicy();
+    if (target_flag & EXCEPTION_REASSEMBLY_MEMCAP)
+        return StreamTcpReassemblyMemcapGetExceptionPolicy();
+    if (target_flag & EXCEPTION_FLOW_MEMCAP)
+        return FlowGetMemcapExceptionPolicy();
+    if (target_flag & EXCEPTION_MIDSTREAM)
+        return StreamMidstreamGetExceptionPolicy();
+    if (target_flag & EXCEPTION_APPLAYER_ERROR)
+        return AppLayerErrorGetExceptionPolicy();
+    return EXCEPTION_POLICY_NOT_SET;
+}
+
 void ExceptionPolicyApply(Packet *p, enum ExceptionPolicy policy, enum PacketDropReason drop_reason)
 {
     SCLogDebug("start: pcap_cnt %" PRIu64 ", policy %u", p->pcap_cnt, policy);
+    if (p->flow) {
+        p->flow->flags |= FLOW_TRIGGERED_EXCEPTION_POLICY;
+        p->flow->applied_exception_policy |= ExceptionPolicyFlag(drop_reason);
```

Comment:
when would we have more than one? Or should this be a simple assignment instead?

In fact, if it is possible that this is already set, what do that mean?

Reply:
> when would we have more than one? Or should this be a simple assignment instead?

Should it maybe be a counter, then?
And then if we have it >1, we check all flags that match? 

> In fact, if it is possible that this is already set, what do that mean?

This did cross my mind, but I wasn't sure this was a reasonable concern, as I didn't recall seeing such a scenario. (although probably because I didn't look for such situations when creating most tests/checks)

Reply:
Could `applied_exception_policy` itself work as the counter, maybe?

Reply:
for now, going with a different approach (inspired by the exception policy counters struct